### PR TITLE
Fixed #26322 -- Updated lazy model reference section of fields docs.

### PR DESCRIPTION
**Pull request by cliff688**
#### Trac ticket number

ticket-26322

#### Branch description
Rewrites the lazy model reference section to list and explain the types of references available.

#### Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: docs/ref/models/fields.txt [hunk 1]
```diff
@@ -1630,95 +1630,30 @@ Django also defines a set of fields that represent relations.
...
 
 Behind the scenes, Django appends ``"_id"`` to the field name to create its
 database column name. In the above example, the database table for the ``Car``
 model will have a ``manufacturer_id`` column. (You can change this explicitly by
-specifying :attr:`~Field.db_column`) However, your code should never have to
+specifying :attr:`~Field.db_column`.) However, your code should never have to
 deal with the database column name, unless you write custom SQL. You'll always
 deal with the field names of your model object.
 
+A database index is automatically created on the ``ForeignKey``. You can
```
Comment (cliff688) [inline docs/ref/models/fields.txt:1651]: Moved this here for better flow in ForeignKey's intro

File: docs/ref/models/fields.txt [hunk 2]
```diff
@@ -2266,6 +2201,113 @@ accepted by :class:`ForeignKey`, plus one extra argument:
...
+        pass
+
+
+    class Car(AbstractCar):
+        pass
+
+
+    # Car.manufacturer will point to `production.Manufacturer` here.
+
+.. admonition:: Blueprint models with relative references
```
Comment (cliff688) [inline docs/ref/models/fields.txt:2285]: Added this as per @charettes's [comment](https://github.com/django/django/pull/6178#discussion_r54976680).

File: docs/ref/models/fields.txt [hunk 3]
```diff
@@ -1630,95 +1630,30 @@ Django also defines a set of fields that represent relations.
...
-consistency rather than joins, or if you will be creating an alternative index
-like a partial or multiple column index.
+See :attr:`ForeignKey.on_delete` for details on the second positional
+argument.
 
 Database Representation
 ~~~~~~~~~~~~~~~~~~~~~~~
 
 Behind the scenes, Django appends ``"_id"`` to the field name to create its
 database column name. In the above example, the database table for the ``Car``
```
Comment (nessita) [inline docs/ref/models/fields.txt:1645]: With the extracted code, this phrase is now confusing, because it references a `Car` class that hasn't been created nor discussed so far. Would you have ideas in how to approach this?

One initial option I could think of is leaving an example in place for using an FK with a `Car` model, but without using any sort of recursion nor lazy reference.

File: docs/ref/models/fields.txt [hunk 4]
```diff
@@ -2266,6 +2201,113 @@ accepted by :class:`ForeignKey`, plus one extra argument:
...
 
+.. _lazy-relationships:
+
+Lazy relationships
+------------------
+
+Lazy relationships allow you to reference models using their names as strings
+or to create recursive relationships. You can use strings as the first
+argument in any relationship field to reference models lazily. The following
+types of lazy references are available:
```
Comment (nessita) [inline docs/ref/models/fields.txt:2212]: In general, I tend to suggest to avoid this type of constructs (using "you"). I think the following is a bit more clear or professional:
```suggestion
Lazy relationships allow referencing models by their names (as strings) or
creating recursive relationships. Strings can be used as the first argument in
any relationship field to reference models lazily.

A lazy references can be either recursive, relative or absolute.
```
(We should definitely add links to each reference type.)

File: docs/ref/models/fields.txt [hunk 5]
```diff
@@ -2266,6 +2201,113 @@ accepted by :class:`ForeignKey`, plus one extra argument:
...
+        manufacturer = models.ForeignKey(
+            "Manufacturer",
+            on_delete=models.CASCADE,
+        )
+        # ...
+
+
+    class Manufacturer(models.Model):
+        # ...
+        pass
```
Comment (nessita) [inline docs/ref/models/fields.txt:2250]: I know this comes from the old content, but I would advice to use the same class definition as the one in `Recursive`, to avoid confusing the reader.

File: docs/ref/models/fields.txt [hunk 6]
```diff
@@ -2266,6 +2201,113 @@ accepted by :class:`ForeignKey`, plus one extra argument:
...
+
+
+    # Car.manufacturer will point to `production.Manufacturer` here.
+
+.. admonition:: Blueprint models with relative references
+
+    Relative references allow you to define reusable blueprint models
+    with abstract classes, as relationships can resolve in concrete
+    subclasses across different applications.
```
Comment (nessita) [inline docs/ref/models/fields.txt:2289]: Thank you! May I suggest:
```suggestion
    Relative references enable the definition of reusable blueprint models with
    abstract classes, allowing relationships to resolve in concrete subclasses
    across different applications.
```

File: docs/ref/models/fields.txt (hunk 2 above)
Comment (nessita) [reply docs/ref/models/fields.txt:2285]: Thank you for working on this @cliff688! For me (non native speaker), reading "blueprint models" raises an eyebrow because I'm not really sure what it means. Perhaps:
```suggestion
.. admonition:: Reusable models with relative references
```

File: docs/ref/models/fields.txt [hunk 7]
```diff
@@ -2266,6 +2201,113 @@ accepted by :class:`ForeignKey`, plus one extra argument:
...
+
+Absolute
+~~~~~~~~
+
+Absolute references allow you to specify a model using its ``app_label``
+and class name, providing a way to reference models across different
+applications. This sort of lazy relationship can be useful in resolving
+circular imports.
```
Comment (nessita) [inline docs/ref/models/fields.txt:2297]: How about:
```suggestion
Absolute references specify a model using its ``app_label`` and class name,
allowing for model references across different applications. This type of lazy
relationship can help resolve circular imports.
```

**Review by nessita — COMMENTED** [review]
@cliff688 Thank you for this work :star2: 

I think this PR it's very well presented and only small tweaks may be needed. I have added comments for your evaluation.

File: docs/ref/models/fields.txt (hunk 3 above)
Comment (cliff688) [reply docs/ref/models/fields.txt]: Good catch! I missed that

File: docs/ref/models/fields.txt (hunk 4 above)
Comment (cliff688) [reply docs/ref/models/fields.txt]: So do I, but because Django docs use the second POV so much (sometimes informally), I am always unsure whether to use the professional tone (third POV), or else to preserve consistency and write in the same tone used in surrounding docs. Maybe we ought to include something in the style guides.

I applied the changes here though.

File: docs/ref/models/fields.txt (hunk 6 above)
Comment (cliff688) [reply docs/ref/models/fields.txt]: I included these changes. Certainly an improvement on what was there before.

But I am not 100% satisfied that we're being as clear as we could be here. My understanding was that the suggestion meant for the docs to point out that because resolutions of relative references in different subclasses can point to different implementations of the pointed-to models, we can use them to create reusable "blueprint" models with relationships.

What do you think?

File: docs/ref/models/fields.txt (hunk 2 above)
Comment (cliff688) [reply docs/ref/models/fields.txt]: Yeah. I think if we can remove or explain "blueprint" in the note itself, then I believe we will have made the meaning clearer. Currently, removing or replacing it with "reusable" does not say anything especially noteworthy of relative references.

Comment (nessita) [reply docs/ref/models/fields.txt]: I see your point @cliff688, thanks for raising that.

One idea would be to provide a small example within this subsection to illustrate this admonition. This would mean that the admonition title gets bumped to a subsubsection (is it possible?) and we provide a concise example. What I don't like about this approach is that this part of the docs is about "references", i.e. study material when following the [Diataxis model](https://diataxis.fr/). According to this, examples and "recipes" should be in the How-To section instead, because those are practical.

So, in summary, I'm not 100% sure of the best course of action. I think that an acceptable middle ground may be the upgrading of the admonition to a entry on its own with the smallest (perhaps slightly incomplete) example along the commentary to explain what a blueprint model is? What do you think?

(In the mid term, one could evaluate a separated, detailed How-To and point to that from here, but I would say that is outside the scope of this ticket.)

Comment (cliff688) [reply docs/ref/models/fields.txt]: Thank you @nessita for link to Diataxis! I enjoyed reading it.

About the admonition, I'm not in favour of adding an example here or in the "How-to" docs. Here's the admonition as it currently stands:
```
.. admonition:: Reusable models with relative references

    Relative references allow the creation of reusable abstract models
    with relationships that can resolve to different implementations
    of the referenced models in various subclasses across different
    applications.
```
Firstly, adding an example might overcomplicate the explanation, especially as the use case is more niche and abstract than common.

Secondly, I think as it currently stands, the admonition is explained in a way that should be clear to the target audience, and for those familiar with the concepts mentioned in the admonition, further illustration should be unnecessary.

Comment (nessita) [reply docs/ref/models/fields.txt]: @cliff688 Then maybe I misunderstood your previous comment. When you said:

> removing or replacing it with "reusable" does not say anything especially noteworthy of relative references.

I thought you were implying that we should add "something" to the current proposal to improve it. Could you expand on what did you mean please?

Comment (cliff688) [reply docs/ref/models/fields.txt]: Oh, I see 👍🏾 When I made that comment, the admonition read:
```
    Relative references enable the definition of reusable blueprint models with
    abstract classes, allowing relationships to resolve in concrete subclasses
    across different applications.
```
which after removing "blueprint" would not say anything that hadn't been said before (i.e. we already mentioned how relative references resolve in concrete model subclasses, and we already know that abstract models are reusable). However, I changed the admonition some moments afterwards (outdating the comment) to:
```
.. admonition:: Reusable models with relative references

    Relative references allow the creation of reusable abstract models
    with relationships that can resolve to different implementations
    of the referenced models in various subclasses across different
    applications.
```
which I think says something noteworthy about relative references in particular; namely, the bit about pointing to different implementations using the same relative reference. Also, with the new wording, we had dropped "blueprint".

Comment (nessita) [reply docs/ref/models/fields.txt]: Amazing, thank you @cliff688. I believe we are in agreement then, we'll keep the admonition as is.

I'll give a final review soon and merge afterwards, great work!

Comment (cliff688) [reply docs/ref/models/fields.txt]: Thank you @nessita 🌻

File: docs/ref/models/fields.txt [hunk 8]
```diff
@@ -1630,95 +1630,33 @@ Django also defines a set of fields that represent relations.
...
-need to use::
+Behind the scenes, Django appends ``"_id"`` to the field name to create its
+database column name. For example, if a foreign key field is defined like::
 
-    class Car(models.Model):
-        manufacturer = models.ForeignKey(
-            "production.Manufacturer",
-            on_delete=models.CASCADE,
-        )
+    manufacturer = models.ForeignKey(Manufacturer, on_delete=models.CASCADE)
```
[the remark this answers is not in the document]
Comment (nessita) [reply docs/ref/models/fields.txt:1647]: Addendum: I worked on this while making my mind on the best solution, so I will push some minor edits.

File: docs/ref/models/fields.txt [hunk 9]
```diff
@@ -1630,95 +1630,33 @@ Django also defines a set of fields that represent relations.
...
 
-This sort of reference, called a lazy relationship, can be useful when
-resolving circular import dependencies between two applications.
+the database table will have a ``manufacturer_id`` column. (You can change
+this explicitly by specifying :attr:`~Field.db_column`.) However, your code
+should never have to deal with the database column name, unless you write
+custom SQL. You'll always deal with the field names of your model object.
 
 A database index is automatically created on the ``ForeignKey``. You can
```
[the remark this answers is not in the document]
Comment (nessita) [reply docs/ref/models/fields.txt:1654]: Addendum: I see from your commit message that this change was intentional. I feel is a bit out of scope from this ticket, so I reverted this bit along with other small changes that I'm pushing.

**Review by nessita — APPROVED** [review]
Thank you @cliff688, while doing the final review, I started making changes locally to help me make better suggestions, so in the end I decided to just push them to not require a third brain context switch for the two of us.

Let me know what you think!

File: docs/ref/models/fields.txt [hunk 10]
```diff
@@ -2266,6 +2211,119 @@ accepted by :class:`ForeignKey`, plus one extra argument:
...
+    from django.db import models
+
+
+    class Car(models.Model):
+        manufacturer = models.ForeignKey(
+            "Manufacturer",
+            on_delete=models.CASCADE,
+        )
+        # ...
+
```
Comment (cliff688) [inline docs/ref/models/fields.txt:2261]: Having introduced a `name` field in the earlier example, I think we should include it here as well, or else replace the `name` with `# ...`, or something like that, as done in other examples

Comment (cliff688) [reply docs/ref/models/fields.txt:2261]: I'm in favour of removing the `name` field for consistency not only with the `Car` model, but also with `Manufacturer` model, which should also have a `name`, but has `# ...` to mean "other fields here"

File: docs/ref/models/fields.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (cliff688) [reply docs/ref/models/fields.txt:1651]: Looks good! Thanks

File: docs/ref/models/fields.txt (hunk 10 above)
Comment (nessita) [reply docs/ref/models/fields.txt:2261]: Sounds good, would you like to fetch the latest and push a new revision with all changes squashed?  Thank you!

File: docs/ref/models/fields.txt [hunk 11]
```diff
@@ -2266,6 +2211,119 @@ accepted by :class:`ForeignKey`, plus one extra argument:
...
+    from django.db import models
+
+
+    class Manufacturer(models.Model):
+        suppliers = models.ManyToManyField("self", symmetrical=False)
+        # ...
+
+When used in an :ref:abstract model <abstract-base-classes>, the recursive
+relationship resolves such that each concrete subclass references itself.
```
Comment (cliff688) [inline docs/ref/models/fields.txt:2242]: @nessita, I reworded this a bit. While more verbose than the previous wording, I believe it's much clearer. Everything else looks great to me. Thank you 🌻

File: docs/ref/models/fields.txt (hunk 10 above)
Comment (nessita) [reply docs/ref/models/fields.txt:2261]: @cliff688 I have reviewed the latest revno available and I still see the `name = ...` definition. I understood your preference was to remove this field, did I misunderstand you? :thinking:

Comment (cliff688) [reply docs/ref/models/fields.txt:2261]: Oh, I think when I was reviewing the changes here in patch form, the `Car` model on `FK` seemed close to the one in Lazy Relationships, so I thought we should keep them consistent, thinking they were in the same context. So, yes, you understood me (thank you), but I had misunderstood the placing of the two `Car` models, so that the issue I had imagined was there disappeared when viewing the file locally as a whole

Comment (cliff688) [reply docs/ref/models/fields.txt:2261]: As for `Manufacturer.name` I think that's harmless.

Comment (nessita) [reply docs/ref/models/fields.txt:2261]: Thank you @cliff688, I pushed final tweaks, I will merge when CI is green again. Amazing work! :trophy:
